libnavigation-core/src/androidTest/java/com/mapbox/navigation/core/trip/service/ArtificialDriverTest.kt

```diff
@@ -0,0 +1,143 @@
+import androidx.test.platform.app.InstrumentationRegistry
+import com.mapbox.navigation.base.options.NavigationOptions
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.MapboxNavigationProvider
+import com.mapbox.navigation.core.navigator.toFixLocation
+import com.mapbox.navigation.core.replay.history.ReplayEventUpdateLocation
+import com.mapbox.navigation.core.replay.history.mapToLocation
+import com.mapbox.navigation.core.replay.route.ReplayRouteMapper
+import com.mapbox.navigation.core.test.R
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigator
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl
+import com.mapbox.navigator.NavigationStatus
+import com.mapbox.navigator.NavigationStatusOrigin
+import com.mapbox.navigator.NavigatorObserver
+import com.mapbox.navigator.RouteState
+import com.mapbox.navigator.SetRoutesReason
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.takeWhile
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.runBlocking
+import org.junit.Assert.assertTrue
+import org.junit.Ignore
+import org.junit.Test
+import org.junit.runner.RunWith
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+@RunWith(AndroidJUnit4::class)
+class ArtificialDriverTest {
+
+    @Test
+    @Ignore("test sometimes fails because of https://mapbox.atlassian.net/browse/NN-418")
+    fun nativeNavigatorFollowsArtificialDriverWithoutReroutes() =
+        runBlocking<Unit>(Dispatchers.Main) {
+            withNavigators { mapboxNavigation, nativeNavigator ->
+                mapboxNavigation.historyRecorder.startRecording()
+                val testRoute = getTestRoute()
+                val events = createArtificialLocationUpdates(testRoute)
+                val setRoutesResult =
+                    nativeNavigator.setRoutes(testRoute, reason = SetRoutesReason.NEW_ROUTE)
+                assertTrue("result is $setRoutesResult", setRoutesResult.isValue)
+                val statusesTracking = async<List<NavigationStatus>> {
+                    nativeNavigator.collectStatuses(untilRouteState = RouteState.COMPLETE)
+                }
+
+                for (location in events.map { it.location.mapToLocation() }) {
+                    assertTrue(nativeNavigator.updateLocation(location.toFixLocation()))
+                }
+
+                val states = statusesTracking.await()
+                val historyFile = suspendCoroutine<String> { continuation ->
+                    mapboxNavigation.historyRecorder.stopRecording {
```

Comment:
There is a suspend version of this method:
```
val historyFile = mapboxNavigation.historyRecorder.startRecording()
```

Reply:
Can you please help me find it? I don't see suspend version [here](https://github.com/mapbox/mapbox-navigation-android/blob/bac2bfbae66d76aecaf22bed999e425f259db54a/libnavigation-core/src/main/java/com/mapbox/navigation/core/history/MapboxHistoryRecorder.kt#L59)

Reply:
You probably meant [this one](https://github.com/mapbox/mapbox-navigation-android/blob/14571b96c14f6d48b5283b8041dae57e8fd11633/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/coroutines/Adapters.kt#L185), but artificial driver test is in a different module as it access internal APIs

Reply:
Yes, this one. Oh, it's a different module. OK then.